pyro/ops/stats.py

```diff
@@ -552,10 +556,43 @@ def energy_score_empirical(pred: torch.Tensor, truth: torch.Tensor) -> torch.Ten
             "Actual shapes: {} versus {}".format(pred.shape, truth.shape)
         )
 
-    retval = (
-        torch.cdist(pred, truth).mean(dim=-2)
-        - 0.5 * torch.cdist(pred, pred).mean(dim=[-1, -2])[..., None]
-    )
+    if pred_batch_size is None:
+        retval = (
+            torch.cdist(pred, truth).mean(dim=-2)
+            - 0.5 * torch.cdist(pred, pred).mean(dim=[-1, -2])[..., None]
+        )
+    else:
+        # Divide predictions into batches
+        pred_len = pred.shape[-2]
+        pred_batches = []
+        while pred.numel() > 0:
+            pred_batches.append(pred[..., :pred_batch_size, :])
+            pred = pred[..., pred_batch_size:, :]
+        # Calculate predictions distance to truth
+        retval = (
+            torch.cat(
+                [
+                    torch.cdist(pred_batch, truth).sum(dim=-2, keepdim=True)
+                    for pred_batch in pred_batches
+                ],
+                dim=-2,
+            ).sum(dim=-2)
```

Comment:
I'm surprised by the `cat(..., dim=-2).sum(dim=-2)` rather than a `sum(....sum(dim=-2))` or `functools.reduce(operator.add, [....sum(dim=-2)])`. If the cat-then-sum is intentional, could you add a comment explaining why you're not reducing early?

Reply:
My initial thought was to use torch native operations, but I didn't stay consistent and used `sum(...)` later on in the function.

After further thought and research I think it would be best to use `stack(...,dim=0).sum(dim=0)` instead of `sum(..)` at both locations, as this allows this function to be JIT compiled, despite being slightly less memory efficient (could also be that `stack(...,dim=0).sum(dim=0)` is more memory efficient if you consider the torch computational tree and gradient calculations).

I have added a commit to that effect and also added a test for JIT compilation.

Nevertheless, I think both options can work so I'm happy to change this if you think otherwise.